cli/cmd/lql.go

```diff
@@ -19,71 +19,215 @@
+		&lqlCmdState.ValidateOnly,
+		"validate_only", "", false,
+		"validate query only (do not run)",
 	)
 }
 
-func runLQLQuery(_ *cobra.Command, args []string) error {
-	var query = ""
+func setQueryFlags(cmds ...*flag.FlagSet) {
+	for _, cmd := range cmds {
+		if cmd != nil {
+			// file flag to specify a query from disk
+			cmd.StringVarP(
+				&lqlCmdState.File,
+				"file", "f", "",
+				"path to an LQL query to run",
+			)
+			// repo flag to specify a query from repo
+			cmd.BoolVarP(
+				&lqlCmdState.Repo,
+				"repo", "r", false,
+				"id of an LQL query to run via active repo",
+			)
+			// url flag to specify a query from url
+			cmd.StringVarP(
+				&lqlCmdState.URL,
+				"url", "u", "",
+				"url to an LQL query to run",
+			)
+		}
+	}
+}
+
+// for commands that take a query as input
+func inputQuery(cmd *cobra.Command, args []string) (
+	query string,
+	err error,
+) {
+	var queryID string
 
+	// if an inline argument was provided
+	// determine if it's a query or a query identifier
 	if len(args) != 0 && args[0] != "" {
-		query = args[0]
-	} else if lqlFile != "" {
-		lqlQuery, err := ioutil.ReadFile(lqlFile)
+		if lqlCmdState.Env || lqlCmdState.Repo {
+			queryID = args[0]
+		} else {
+			query = args[0]
+		}
+	}
+
+	if lqlCmdState.Env {
+		var queryResponse api.LQLQueryResponse
+		queryResponse, err = cli.LwApi.LQL.GetQueryByID(queryID)
+		if err == nil && len(queryResponse.Data) != 0 {
+			query = queryResponse.Data[0].QueryText
+		}
+	} else if lqlCmdState.Repo {
+		err = errors.New("NotImplementedError")
+	} else if lqlCmdState.File != "" {
+		var fileData []byte
+		fileData, err = ioutil.ReadFile(lqlCmdState.File)
 		if err != nil {
-			return errors.Wrap(err, "unable to read file")
+			err = errors.Wrap(err, "unable to read file")
+			return
 		}
-		query = string(lqlQuery)
+		query = string(fileData)
+	} else if lqlCmdState.URL != "" {
+		msg := "unable to open URL"
+		var response *http.Response
+		var body []byte
+
+		response, err = http.Get(lqlCmdState.URL)
+		if err != nil {
+			err = errors.Wrap(err, msg)
+			return
+		}
+		defer response.Body.Close()
+
+		if response.StatusCode != 200 {
+			err = errors.Wrap(errors.New(response.Status), msg)
+			return
+		}
+
+		body, err = ioutil.ReadAll(response.Body)
+		if err != nil {
+			err = errors.Wrap(err, msg)
+			return
+		}
+		query = string(body)
 	} else {
-		// avoid asking for a confirmation before launching the editor
+		var firstUseWord string
+		if lqlCmdState.ValidateOnly {
+			firstUseWord = "validate"
+		} else {
+			firstUseWord = strings.Split(cmd.Use, " ")[0]
+		}
 		prompt := &survey.Editor{
-			Message:  "Type an LQL query to run",
+			Message:  fmt.Sprintf("Type a query to %s", firstUseWord),
 			FileName: "query*.sh",
 		}
-		err := survey.AskOne(prompt, &query)
-		if err != nil {
-			return err
-		}
+		err = survey.AskOne(prompt, &query)
 	}
 
-	cli.Log.Debugw("running LQL query", "query", query)
-	response, err := cli.LwApi.LQL.Query(query)
+	return
+}
```

Comment:
The logic in this function has to be reduced. I will review it once the other comments are resolved, specially the one that tries to remove the `--env` flag.

Reply:
Fixed.